### PR TITLE
#168781808 implement user password reset

### PR DESCRIPTION
**Pull request by swaibat**
#### What does this PR do?
- Implements user password reset incase a user forgot his password, where a notification is sent to user email for reset instructions
#### Description of Task to be completed?
- validate input email address
- install `nodemailer` package
- add new data in `.env sample`
- implement email sender
- validate password on reset
- confirm password
#### How should this be manually tested?
- `npm i `to install relevant package added `nodemail`
- add new environments in `.env file`
```
MAIL_HOST:                 #[like gmail.com]
MAIL_PORT:                 #[any port can be null or excluded]
MAIL_SENDER:               #[any sender barfoot nomad eagels]
MAIL_USER:                 #[any email sender example@gmail.com]
MAIL_PASSWORD:             #[any email sender password]
MAIL_SERVICE:              #[like gmail]

```
- add new packages
- access `hostaddress/api/v1/users/reset-password`
- provide the email to be reset in the body.
- notification and instructions are sent to the email provided.
- a link redirects you requests to add `new password` and confirm it
#### Any background context you want to provide?
**Basic analysis design**
Screen Shot 2019-10-12 at 09 24 05

#### What are the relevant pivotal tracker 
[#168781808](https://www.pivotaltracker.com/story/show/168781808)
#### Screenshots (if appropriate)
Screen Shot 2019-10-10 at 17 04 30

Screen Shot 2019-10-10 at 16 58 27
Screen Shot 2019-10-11 at 12 10 16

#### Questions:
na

### CODE REVIEW

**Review by Crispy-rw — APPROVED** [review]

**Review by lemoissonM — APPROVED** [review]

**Review by alexismajya — APPROVED** [review]

**Review by urbanishimwe — APPROVED** [review]

**Review by urbanishimwe — DISMISSED** [review]
there is conflicts!

File: src/controllers/email.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import helpers from '../utils/helper';
+import db from '../database/models/index';
+import sendResult from '../utils/sendResult';
+import { msg, transporter } from '../config';
+
+const email = {
+  async sendReset(req, res) {
```
Comment (nakiwuge) [inline src/controllers/email.js]: why are you using async here?

File: src/controllers/email.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    const message = { ...msg(req, helpers), to: req.user.email };
+    transporter.sendMail(message, () => sendResult(res, 201, `password reset instructions sent to ${req.user.email}`));
+  },
```
Comment (nakiwuge) [inline src/controllers/email.js:13]: add a new line after this line

File: src/controllers/email.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  async resetPass(req, res) {
+    const userInfo = await helpers.decodeToken(req.params.token);
+    if (userInfo.error) return sendResult(res, 400, userInfo.error);
```
Comment (nakiwuge) [inline src/controllers/email.js:17]: Add a newline here

File: src/middlewares/userMiddlware.js [hunk 4]
```diff
@@ -13,6 +14,25 @@ const User = {
       return sendResult(res, 400, 'Both email and password are required');
     }
     next();
-  }
+  },
```
Comment (nakiwuge) [inline src/middlewares/userMiddlware.js:18]: new line here

File: src/middlewares/userMiddlware.js [hunk 5]
```diff
@@ -13,6 +14,25 @@ const User = {
...
+  validateEmail(req, res, next) {
+    const validateEmail = new Check({ email: req }).req().email().error;
+    validateEmail
+      ? sendResult(res, 400, validateEmail)
+      : next();
+  },
```
Comment (nakiwuge) [inline src/middlewares/userMiddlware.js:25]: Add anew line here

File: src/utils/helper.js [hunk 6]
```diff
@@ -1,6 +1,7 @@
 /* eslint-disable require-jsdoc */
 import jwt from 'jsonwebtoken';
 import bcrypt from 'bcrypt';
+import '@babel/polyfill';
```
Comment (nakiwuge) [inline src/utils/helper.js]: why are you importing bablel here?

**Review by nakiwuge — CHANGES_REQUESTED** [review]
Good work @swaibat however currently I can change reset my password with a non-alphanumeric password. Also, you have more than 3 commits kindly squash them as we had suggested.
I have also left other comments on the pr that you should check out.

**Review by urbanishimwe — APPROVED** [review]

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**